src/sagemaker/s3_utils.py

```diff
@@ -0,0 +1,118 @@
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+#     http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+"""This module contains helper functions related to S3. You may want to use `s3.py` instead.
+
+This has a subset of the functions available through s3.py. This module was initially created with
+functions that were originally in `s3.py` so that those functions could be imported inside
+`session.py` without circular dependencies. (`s3.py` imports Session as a dependency.)
+"""
+from __future__ import print_function, absolute_import
+
+import pathlib
+import logging
+from typing import Optional
+
+from six.moves.urllib.parse import urlparse
+
+logger = logging.getLogger("sagemaker")
+
+
+def parse_s3_url(url):
+    """Returns an (s3 bucket, key name/prefix) tuple from a url with an s3 scheme.
+
+    Args:
+        url (str):
+
+    Returns:
+        tuple: A tuple containing:
+
+            - str: S3 bucket name
+            - str: S3 key
+    """
+    parsed_url = urlparse(url)
+    if parsed_url.scheme != "s3":
+        raise ValueError("Expecting 's3' scheme, got: {} in {}.".format(parsed_url.scheme, url))
+    return parsed_url.netloc, parsed_url.path.lstrip("/")
+
+
+def s3_path_join(*args):
+    """Returns the arguments joined by a slash ("/"), similar to ``os.path.join()`` (on Unix).
+
+    If the first argument is "s3://", then that is preserved.
+
+    Args:
+        *args: The strings to join with a slash. None or empty strings are skipped. For example,
+            an args of ["", "key1", None, "key2"] would yield an output of "key1/key2".
+
+    Returns:
+        str: The joined string, without a slash at the end.
+    """
+    filtered_args = list(filter(lambda item: item is not None and item != "", args))
+    if len(filtered_args) == 0:
+        return ""
+
+    if filtered_args[0].startswith("s3://"):
+        path = str(pathlib.PurePosixPath(*filtered_args[1:])).lstrip("/")
+        return str(pathlib.PurePosixPath(filtered_args[0], path)).replace("s3:/", "s3://")
+
+    return str(pathlib.PurePosixPath(*filtered_args)).lstrip("/")
+
+
+def s3_path_join_with_end_slash(*args):
+    """Deprecated, use :func:`s3_path_join`. This returns the same output, except with an end slash.
+
+    This function was added to maintain backwards compatibility because some S3 paths explicitly
+    had an end slash added. It may be the case that this end slash is not actually necessary to
+    maintain, in which case this function can be removed.
+    """
+    path = s3_path_join(*args)
+
+    if path:
+        return path + "/"
+
+    return path
+
+
+def calculate_bucket_and_prefix(
```

Comment:
nit: can we use naming as determine_bucket_and_prefix/derive_bucket_and_prefix ?

Reply:
these are much better names! thanks!